infra/gravity/terraform.go

```diff
@@ -146,8 +148,105 @@ func saveResourceAllocations() error {
 	return nil
 }
 
+// makeDynamicParams takes base config, validates it and returns cloudDynamicParams
+func makeDynamicParams(baseConfig ProvisionerConfig) (*cloudDynamicParams, error) {
+	param := cloudDynamicParams{ProvisionerConfig: baseConfig}
+
+	// OS name is cloud-init script specific
+	// enforce compatible values
+	var ok bool
+	usernames := map[string]map[string]string{
+		"azure": map[string]string{
+			"ubuntu": "robotest",
+			"debian": "admin",
+			"redhat": "redhat", // TODO: check
+			"centos": "centos",
+			"suse":   "robotest",
+		},
+		"aws": map[string]string{
+			"ubuntu": "ubuntu",
+			"debian": "admin",
+			"redhat": "redhat",
+			"centos": "centos",
+		},
+	}
+
+	param.user, ok = usernames[baseConfig.CloudProvider][baseConfig.os]
+	if !ok {
+		return nil, trace.BadParameter("%s ")
+	}
+
+	param.homeDir = filepath.Join("/home", param.user)
+
+	param.tf = terraform.Config{
+		CloudProvider: baseConfig.CloudProvider,
+		ScriptPath:    baseConfig.ScriptPath,
+		NumNodes:      int(baseConfig.nodeCount),
+		OS:            baseConfig.os,
+	}
+
+	if baseConfig.AWS != nil {
+		aws := *baseConfig.AWS
+		param.tf.AWS = &aws
+		param.tf.AWS.ClusterName = baseConfig.tag
+		param.tf.AWS.SSHUser = param.user
+
+		param.env = map[string]string{
+			"AWS_ACCESS_KEY_ID":     param.tf.AWS.AccessKey,
+			"AWS_SECRET_ACCESS_KEY": param.tf.AWS.SecretKey,
+			"AWS_DEFAULT_REGION":    param.tf.AWS.Region,
+		}
+	}
+
+	if baseConfig.Azure != nil {
+		azure := *baseConfig.Azure
+		param.tf.Azure = &azure
+		param.tf.Azure.ResourceGroup = baseConfig.tag
+		param.tf.Azure.SSHUser = param.user
+	}
+
+	return &param, nil
+}
+
+func runTerraform(ctx context.Context, baseConfig ProvisionerConfig, logger logrus.FieldLogger) (nodes []infra.Node, destroyFn func(context.Context) error, params *cloudDynamicParams, err error) {
+	retr := wait.Retryer{
+		Delay:       defaults.TerraformRetryDelay,
+		Attempts:    defaults.TerraformRetries,
+		FieldLogger: logger,
+	}
+
+	retry := 0
+	cfg := baseConfig
+
+	err = retr.Do(ctx, func() error {
+		if retry != 0 {
+			cfg = baseConfig.WithTag(fmt.Sprintf("R%d", retry))
+			logger.Info("retrying terraform provisioning")
+		}
+		retry++
+
+		params, err = makeDynamicParams(cfg)
```

Comment:
Why would this fail after the first call? I mean, why not `makeDynamicParams` outside of the retry loop - does the `cfg` change?

Reply:
yes, three lines above